### PR TITLE
Extend RedisSettings to include redis Retry Helper settings

### PR DESCRIPTION
**Pull request by mernmic**
proposed fix for https://github.com/samuelcolvin/arq/issues/386

### CODE REVIEW

File: arq/connections.py
Comment (JonasKs) [inline arq/connections.py:52]: This isn't an `Any`, but rather a `redis.Retry`, right?

**Review by JonasKs — CHANGES_REQUESTED** [review]

File: arq/connections.py
Comment (mernmic) [reply arq/connections.py]: yup, fixed.

**Comment by JonasKs** [comment]
The tests has been fixed in #382, so don’t worry about those failing pipelines on some Python versions for now.  

As you can see from the codecov bot above, this repo strive to have 100% codecov. Would love if you added tests for these settings.   
If you’re unsure how to proceed, please don’t hesitate to ask. 😊

**Comment by mernmic** [comment]
@JonasKs, added some tests which improved coverage.  Still one partial, but might be from testing failures?

**Comment by JonasKs** [comment]
I'll try to review again this weekend / early next week 🙇‍♂️

File: arq/connections.py
Comment (JonasKs) [inline arq/connections.py:52]: Wouldn't this be `Optional[List[Exception]]`?

**Review by JonasKs — APPROVED** [review]
Looks good to me! 😊

File: arq/connections.py
Comment (mernmic) [reply arq/connections.py]: Yup, fixed.  thanks for catching.

**Comment by mernmic** [comment]
> Looks good to me! 😊

Thanks for taking a look., just pushed a change for the Exception type.

We've been using the above changes in our systems and everything has been working as expected so far.

A heads up I'll be offline from the first few weeks of February, so if anything comes up that's the reason for any radio silence on my end.

Thanks again!

**Comment by JonasKs** [comment]
Thanks for the PR and for testing in your environment! 
Enjoy your offline time☺️     

I'd also recommend double checking that maintainers can edit your PR, in case Samuel finds time to review/release while you're gone. 😊

**Comment by maximgrigorov** [comment]
Hi there,

is it possible to merge this fix to main branch?

**Comment by akira** [comment]
Hello- thanks for the fix! Any thoughts on merging this? Would be great to have for my use cases! 🙇

**Comment by mernmic** [comment]
Fix while waiting for PR merge.

Initialize ArqRedis with a connection pool including your retry helper.

```
class WorkerSettingsBase:
    redis_pool = ArqRedis(
        connection_pool=redis.asyncio.ConnectionPool.from_url(
            url=RedisSecrets.REDIS_URL,
            retry_on_timeout=True,
            retry_on_error=[
                redis.asyncio.TimeoutError,
                redis.asyncio.ConnectionError,
            ],
            retry=Retry(
                backoff=CustomRedisBackoff(),
                retries=5,
            )
        )
    ),
   ....

if __name__ == "__main__":
    arq.run_worker(WorkerSettingsBase)
```

redis_pool parameter can be found in docs here:
https://arq-docs.helpmanual.io/#arq.worker.Worker

Example custom backoff class, we are using for more granular access and logging but you can use any redis retry helper
```
class CustomRedisBackoff(AbstractBackoff):
    def __init__(self, cap=0.512, base=0.008):
        """
        `cap`: maximum backoff time in seconds
        `base`: base backoff time in seconds
        """
        self.backoff = ExponentialBackoff(cap=cap, base=base)

    def compute(self, failures):
        sleep = self.backoff.compute(failures)
        logger.info(
            {
                "message": "redis connection error, triggering backoff",
                "failures": failures,
                "backoff": repr(self.backoff),
                "sleep (secs)": sleep,
            }
        )
        return sleep
```

**Review by samuelcolvin — APPROVED** [review]
Thanks so much for this, sorry for the long delay.

**Comment by samuelcolvin** [comment]
@JonasKs @Kludex we might be able to use redis's retry logic to remove our retry stuff - https://github.com/samuelcolvin/arq/blob/94cd8782b4f0764a17962186a349d32125cb98e3/arq/connections.py#L258-L286

**Comment by samuelcolvin** [comment]
`v0.26.0b1` is released, please try it, I'll release `v0.26` at the end of the week, see #441.

**Pull request merged**
